go/extractor/cli/go-autobuilder/go-autobuilder.go

```diff
@@ -692,3 +695,161 @@ func main() {
 }
+
+const minGoVersion = "1.11"
+const maxGoVersion = "1.20"
+
+func outsideSupportedRange(version string) bool {
+	short := semver.MajorMinor("v" + version)
+	return semver.Compare(short, "v"+minGoVersion) < 0 || semver.Compare(short, "v"+maxGoVersion) > 0
+}
+
+func checkForUnsupportedVersions(v versionInfo) (msg, version string) {
+	if v.goDirectiveFound && outsideSupportedRange(v.goModVersion) {
+		msg = "The version of Go found in the `go.mod` file (" + v.goModVersion +
+			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion + ")."
+		version = ""
+		diagnostics.EmitUnsupportedVersionGoMod(msg)
+	}
+
+	if v.goInstallationFound && outsideSupportedRange(v.goEnvVersion) {
+		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
+			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion + ")."
+		version = ""
+		diagnostics.EmitUnsupportedVersionEnvironment(msg)
+	}
+
+	return msg, version
+}
+
+func checkForVersionsNotFound(v versionInfo) (msg, version string) {
+	if !v.goInstallationFound && !v.goDirectiveFound {
+		msg = "No version of Go installed and no `go.mod` file found. Writing an " +
+			"`environment.json` file specifying the maximum supported version of Go (" +
+			maxGoVersion + ")."
+		version = maxGoVersion
+		diagnostics.EmitNoGoModAndNoGoEnv(msg)
+	}
+
+	if !v.goInstallationFound && v.goDirectiveFound {
+		msg = "No version of Go installed. Writing an `environment.json` file specifying the " +
+			"version of Go found in the `go.mod` file (" + v.goModVersion + ")."
+		version = v.goModVersion
+		diagnostics.EmitNoGoEnv(msg)
+	}
+
+	if v.goInstallationFound && !v.goDirectiveFound {
+		msg = "No `go.mod` file found. Version " + v.goEnvVersion + " installed in the environment."
+		version = ""
+		diagnostics.EmitNoGoMod(msg)
+	}
+
+	return msg, version
+}
+
+func compareVersions(v versionInfo) (msg, version string) {
+	if semver.Compare("v"+v.goModVersion, "v"+v.goEnvVersion) > 0 {
+		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
+			") is lower than the version found in the `go.mod` file (" + v.goModVersion +
+			").\nWriting an `environment.json` file specifying the version of Go from the " +
+			"`go.mod` file (" + v.goModVersion + ")."
+		version = v.goModVersion
+		diagnostics.EmitVersionGoModHigherVersionEnvironment(msg)
+	} else {
+		msg = "The version of Go installed in the environment (" + v.goEnvVersion +
+			") is high enough for the version found in the `go.mod` file (" + v.goModVersion + ")."
+		version = ""
+		diagnostics.EmitVersionGoModNotHigherVersionEnvironment(msg)
+	}
+
+	return msg, version
+}
+
+func getVersionToInstall(v versionInfo) (msg, version string) {
+	msg, version = checkForUnsupportedVersions(v)
+	if msg != "" {
+		return msg, version
+	}
+
+	msg, version = checkForVersionsNotFound(v)
+	if msg != "" {
+		return msg, version
+	}
+
+	msg, version = compareVersions(v)
+	return msg, version
+}
```

Comment:
The logic implemented in these functions is relatively complicated and tricky to follow / it is easy to miss corner cases. Once tests are added, those will increase confidence, but it would also be good to add documentation and more explicitly document assumptions/invariants throughout these functions.

Reply:
I'll add docstrings to the functions.